contracts/Items/DEFYForge.sol

```diff
@@ -0,0 +1,241 @@
+        );
+
+        // require input arrays are not null
+        require(
+            inputMaterialsAmount.length != 0 && inputMaterialsId.length != 0,
+            "DEFYForge: Invalid input materials"
+        );
+
+        // require input materials ids and amount are the same
+        require(
+            inputMaterialsAmount.length == inputMaterialsId.length,
+            "DEFYForge: All arrays must be the same length"
+        );
+
+        // require operative has tokens available
+        for (uint256 i = 0; i < inputMaterialsId.length; i++) {
+            require(
+                itemContract.balanceOf(operativeAddress, inputMaterialsId[i]) >=
+                    inputMaterialsAmount[i],
+                "DEFYForge: Operative does not have suffient materials"
+            );
+        }
+
+        // create print job record
+        bytes32 printJobId = computeNextPrintJobIdForOperative(
+            operativeAddress
+        );
+
+        printJobs[printJobId] = PrintJob(
+            itemContract,
+            operativeAddress,
+            inputMaterialsId,
+            inputMaterialsAmount,
+            blueprintId,
+            outputPartId,
+            false
+        );
+
+        printJobIds.push(printJobId);
+        uint256 currentPrintJobCount = operativesPrintJobCount[
+            operativeAddress
+        ];
+        operativesPrintJobCount[operativeAddress] = currentPrintJobCount + 1;
+
+        // burn input materials
+
+        for (uint256 i = 0; i < inputMaterialsId.length; i++) {
+            itemContract.burnToken(
+                operativeAddress,
+                inputMaterialsId[i],
+                inputMaterialsAmount[i]
+            );
+        }
+
+        // return printJobId
+        return printJobId;
+    }
+
+    // pass in printJobId and mint forged part
+    function completeForge(bytes32 _printJobId)
+        public
+        onlyRole(FORGER_ROLE)
+        returns (address, uint256)
+    {
+        // require printJob has not been completed
+        require(
+            getPrintJob(_printJobId).completed == false,
+            "DEFYForge: Forge has been completed"
+        );
+
+        // get PrintJob struct
+        PrintJob memory _printJob = getPrintJob(_printJobId);
+        bytes memory zeroBytes;
+
+        // Mint output part
+        _printJob.itemContract.mint(
+            _printJob.operativeAddress,
+            _printJob.outputPartId,
+            1,
+            zeroBytes
+        );
+
+        printJobs[_printJobId].completed = true;
+
+        return (_printJob.operativeAddress, _printJob.outputPartId);
+    }
+
+    /**
+     * @dev Returns the print job id at the given index.
+     * @return the print job id
+     */
+    function getPrintJobIdAtIndex(uint256 index)
+        external
+        view
+        returns (bytes32)
+    {
+        require(index < getPrintJobsCount(), "DEFYForge: index out of bounds");
+        return printJobIds[index];
+    }
```

Comment:
No need for this, as the print jobs will be directly sequential.  Instead it should be `forgeJobIdAtIndexForOperative`